crates/sui-core/src/accumulators/mod.rs

```diff
@@ -339,34 +342,126 @@ impl AccumulatorSettlementTxBuilder {
-    ) -> Vec<TransactionKind> {
+    ) -> (
+        Vec<TransactionKind>, /* settlements */
+        TransactionKind,      /* barrier */
+    ) {
+        let Self { updates, addresses } = self;
+
+        let mut pending_updates = Vec::new();
+        let mut settlements = Vec::new();
+
+        let build_one_settlement_txn = |idx: u64, updates: &mut Vec<(AccumulatorObjId, Update)>| {
+            let (total_input_sui, total_output_sui) =
+                updates
+                    .iter()
+                    .fold((0, 0), |(acc_input, acc_output), (_, update)| {
+                        (acc_input + update.input_sui, acc_output + update.output_sui)
+                    });
+
+            Self::build_one_settlement_txn(
+                &addresses,
+                epoch,
+                idx,
+                checkpoint_height,
+                accumulator_root_obj_initial_shared_version,
+                updates.drain(..),
+                total_input_sui,
+                total_output_sui,
+            )
+        };
+
+        for (obj, update) in updates.into_iter() {
+            pending_updates.push((obj, update));
+
+            if pending_updates.len() == protocol_config.max_updates_per_settlement_txn() as usize {
+                settlements.push(build_one_settlement_txn(
+                    settlements.len() as u64,
+                    // pending_updates will be drained and can be re-used
+                    &mut pending_updates,
+                ));
+            }
+        }
+
+        if !pending_updates.is_empty() {
+            settlements.push(build_one_settlement_txn(
+                settlements.len() as u64,
+                &mut pending_updates,
+            ));
+        }
+
+        // Now construct the barrier transaction
+        let mut builder = ProgrammableTransactionBuilder::new();
+        let root = builder
+            .input(CallArg::Object(ObjectArg::SharedObjectV2 {
+                id: SUI_ACCUMULATOR_ROOT_OBJECT_ID,
+                initial_shared_version: accumulator_root_obj_initial_shared_version,
+                mutability: SharedObjectMutability::Mutable,
+            }))
+            .unwrap();
+
+        let epoch_arg = builder.pure(epoch).unwrap();
+        let checkpoint_height_arg = builder.pure(checkpoint_height).unwrap();
+        let idx_arg = builder.pure(settlements.len() as u64).unwrap();
+        let total_input_sui_arg = builder.pure(0u64).unwrap();
+        let total_output_sui_arg = builder.pure(0u64).unwrap();
+
+        builder.programmable_move_call(
+            SUI_FRAMEWORK_PACKAGE_ID,
+            ACCUMULATOR_SETTLEMENT_MODULE.into(),
+            ACCUMULATOR_ROOT_SETTLEMENT_PROLOGUE_FUNC.into(),
+            vec![],
+            vec![
+                root,
+                epoch_arg,
+                checkpoint_height_arg,
+                idx_arg,
+                total_input_sui_arg,
+                total_output_sui_arg,
+            ],
+        );
```

Comment:
I wonder if it is cleaner to merge this part of code with build_one_settlement_txn since you are calling the same move function anyway. And it takes an enum mode where either it's settlement mode with updates or it's barrier mode.